src/block-ref.test.ts

```diff
@@ -1,59 +1,620 @@
+}
+
+/**
+ * Options supported by `withTestSetup`.
+ *
+ * @property configureMiddleware - A function which determines which middleware
+ * should be added to the engine.
+ */
+interface WithTestSetupOptions {
+  configureMiddleware: (setup: Setup) => {
+    middlewareUnderTest: JsonRpcMiddleware<any, any>;
+    otherMiddleware?: JsonRpcMiddleware<any, any>[];
+  };
+}
+
+/**
+ * The function that `withTestSetup` is expected to take and will call once the
+ * setup objects are created.
+ *
+ * @template T - The type that the function will return, minus the promise
+ * wrapper.
+ */
+type WithTestSetupCallback<T> = (setup: Setup) => Promise<T>;
+
+/**
+ * An object that can be used to assign a canned response to a request made via
+ * `provider.sendAsync`.
+ *
+ * @template T - The type that represents the request params.
+ * @template U - The type that represents the response result.
+ * @property request - An object that represents a JsonRpcRequest. Keys such as
+ * `id` or `jsonrpc` may be omitted if you don't care about them.
+ * @property response - A function that returns a JsonRpcResponse for that
+ * request. This function takes two arguments: the *real* request and a
+ * `callNumber`, which is the number of times the request has been made
+ * (counting the first request as 1). This latter argument be used to specify
+ * different responses for different instances of the same request.
+ * @property remainAfterUse - Usually, when a request is made via
+ * `provider.sendAsync`, the ProviderRequestStub which matches that request is
+ * removed from the list of stubs, so that if the same request comes through
+ * again, there will be no matching stub and an error will be thrown. This
+ * feature is useful for making sure that all requests have canned responses.
+ */
+interface ProviderRequestStub<T, U> {
+  request: Partial<JsonRpcRequest<T>>;
+  response: (
+    request: JsonRpcRequest<T>,
+    callNumber: number,
+  ) => JsonRpcResponse<U>;
+  remainAfterUse?: boolean;
+}
+
+describe('createBlockRefMiddleware', () => {
+  // This list corresponds to the list in the `blockTagParamIndex` function
+  // within `src/utils/cache.ts`
+  (
+    [
+      { blockParamIndex: 0, methods: ['eth_getBlockByNumber'] },
+      {
+        blockParamIndex: 1,
+        methods: [
+          'eth_getBalance',
+          'eth_getCode',
+          'eth_getTransactionCount',
+          'eth_call',
+        ],
+      },
+      { blockParamIndex: 2, methods: ['eth_getStorageAt'] },
+    ] as const
+  ).forEach(({ blockParamIndex, methods }) => {
+    methods.forEach((method: string) => {
+      describe(`when the RPC method is ${method}`, () => {
+        describe('if the block param is "latest"', () => {
+          it('makes a direct request through the provider, replacing the block param with the latest block number', async () => {
+            await withTestSetup(
+              {
+                configureMiddleware: ({ provider, blockTracker }) => {
```

Comment:
This is a slightly different style compared with the `retryOnEmpty` tests I wrote earlier. I wanted to make it so that:

* It's obvious what the middleware-under-test is. Previously this was automatically added in a helper function, now it happens within each test explicitly. Yet, it's drawn out so it's easier to see.
* If you don't need to supply a final middleware (because you're not testing that the middleware-under-test is calling the next middleware), you don't need to supply an extra option like `withFinalMiddleware: false`. At the same time, a default middleware is still pushed onto the engine. The purpose of this default middleware is to serve as a reminder for the developer in case the middleware-under-test attempts to call the next middleware and that was unexpected.
* If you do need to supply a final middleware because you want to test that it gets called, then you can still do that easily.